### ISSUE TITLE
Why doesn't the CLI configure API access for my authenticated users?

### ISSUE THREAD

**Issue by orome** [aws-amplify/amplify-category-api#1326]

**Comment by orome** [comment]
@josefaidt How does that interact with the instructions I was given by Support:
> 1. Create a Cognito user pool authorizer in API Gateway console. The token source value is "Authorization"
> 2. Add the Cognito user pool authorizer to ANY method of the /charge resource path
> 3. Deploy API

**Comment by orome** [comment]
@josefaidt Following those instructions, when I push:

1. Two additional files are added under `amplify/backend/`: `tsconfig.json` and `package.json`
2. I get: 
    `Building resource api/Test🛑 Failed to override Test due to: InvalidOverrideError: Packaging overrides failed..`

**Comment by josefaidt** [comment]
Hey @orome I would recommend the route with resource overrides as this will ensure your modifications are applied across subsequent updates with `amplify update api` and `amplify push`

To your second note, did you run `amplify override api` and select REST API?

**Comment by orome** [comment]
@josefaidt 
> To your second note, did you run `amplify override api` and select REST API?

Yes. In any case, it works!

So just to clarify:

1. What exactly has this done? It didn't do anything in AWS Gateway (that's nice and clean): it's still IAM there.
2. So I assume that outside my app, IAM is still required (also nice and clean)(?
3. If so, how do I authenticate access from Postman or curl? What does the header look like and where do I get the token I need to pass?

**Comment by josefaidt** [comment]
> What exactly has this done? It didn't do anything in AWS Gateway (that's nice and clean): it's still IAM there.

You will want to step through `amplify update api` and remove the restrictions, then apply the override with the linked Cognito example

> So I assume that outside my app, IAM is still required (also nice and clean)(?

Yes, and once the override is applied the routes will need the JWT (shown in your example above)

> If so, how do I authenticate access from Postman or curl? What does the header look like and where do I get the token I need to pass?

You will want to pass an authorization header similar to the code example from your original post, for example:

```tsx
import useSWR from 'swr'
import { Auth, API } from 'aws-amplify'
import '@aws-amplify/ui-react/styles.css'
import awsExports from '../aws-exports'

const [{ name: API_NAME }] = awsExports.aws_cloud_logic_custom

async function fetcher(route) {
  const response = await API.get(API_NAME, route, {
    headers: {
      Authorization: `Bearer ${(await Auth.currentSession())
        .getIdToken()
        .getJwtToken()}`,
    },
  })
  return response
}

export default function HomePage(props) {
  const { data, error } = useSWR('/hello', fetcher)

  return (
    <div>
      {!data && !error && <p>Loading...</p>}
      {error && <p>Error: {error.message}</p>}
      {data && (
        <pre>
          <code>{JSON.stringify(data, null, 2)}</code>
        </pre>
      )}
    </div>
  )
}
```

**Comment by orome** [comment]
@josefaidt Sorry I wasn't clear: I completely deleted the original API and thus started fresh with your suggestion, which works great (no need to pass the Authorization in my code at all. So my follow ups are:

1. Am I correct in assuming (since the API Gateway itself is still limited to IAM) that the "override" is essentially creating and providing its own IAM credentials for use by the app?
2. What header and value should I provide for use outside the app? For example in Postman or using curl? Do I need to use one of my existing IAM credentials (or create one), or is there a way to get access to the IAM credentials that the app is using (if 1 is correct)?

**Comment by josefaidt** [comment]
> Am I correct in assuming (since the API Gateway itself is still limited to IAM) that the "override" is essentially creating and providing its own IAM credentials for use by the app?

The prompt from the CLI to "restrict access" will use IAM, however when using the override it will use Cognito User Pools in place, meaning any call to the REST API will need to contain the caller's token

> What header and value should I provide for use outside the app? For example in Postman or using curl? Do I need to use one of my existing IAM credentials (or create one), or is there a way to get access to the IAM credentials that the app is using (if 1 is correct)?

If you use the override linked earlier in the thread this will be the token output from:

```js
Authorization: `Bearer ${(await Auth.currentSession()).getIdToken().getJwtToken()}`,
```

If you would like to test your REST API with IAM you can either strip the authorization header out of a request from your app, or use Postman's built-in functionality to construct the signature.

This example is in a different context, but serves as a point of reference for constructing the IAM auth header https://docs.amplify.aws/lib/graphqlapi/graphql-from-nodejs/q/platform/js/#iam-authorization

**Comment by orome** [comment]
> ... however when using the override it will use Cognito User Pools in place, meaning any call to the REST API will need to contain the caller's token.

So where is that code running? Where do I see it in the console? API Gateway still says IAM.

**Comment by josefaidt** [comment]
You will need to remove the IAM restrictions before applying the override with `amplify update api`, then use `amplify override api` and select `REST` to create the `override.ts` file. Then use [this guide](https://docs.amplify.aws/cli/restapi/override/) to create the relationship between your Cognito User Pools resource and the API Gateway resource

**Comment by josefaidt** [comment]
Yes, remove the access restrictions with `amplify update api` to remove IAM. Since we're not creating a Cognito authorizer it may not display as "cognito" in the same section you see IAM in, but it will be authenticated

**Comment by orome** [comment]
@josefaidt If I remove the resurgent for authenticated users for an existing and previously restricted patth (here `/test`) using `amplify update api` with
```text
✔ What would you like to do? · Update path
? Select the path to edit /test
✔ Provide a path (e.g., /book/{isbn}): · /test
✔ Choose a Lambda source · Use a Lambda function already added in the current Amplify project
✔ Choose the Lambda function to invoke by this path · LearningChat01
✔ Restrict API access? (Y/n) · no
```
it still says IAM in API Gateway.

**Comment by orome** [comment]
@josefaidt Actually what [you say above](https://github.com/aws-amplify/amplify-category-api/issues/1326#issuecomment-1458684150) is wrong. IAM needs to be there for the Cognito override to work.

**Comment by orome** [comment]
@josefaidt Actually what [you say above](https://github.com/aws-amplify/amplify-category-api/issues/1326#issuecomment-1458684150) is wrong. IAM needs to be there for the Cognito override to work.

**Comment by orome** [comment]
So it's:

1. In `update`/`add`**limit access to authenticated users** ~~don't limit access at all~~.
2. Use `override` as described to require Cognito authentication

And if I do that it will say **IAM** ~~Cognito~~ in API Gateway. There's no way to get rid of IAM in API Gateway, it seems.

So the original question is back: Does access from outside my app use IAM (as indicated). Where is the override code running? Does it intercept all calls to the REST endpoints (not just from my app)?

**Comment by orome** [comment]
@josefaidt Actually I'm no longer sure what's going on, push seems no longer to work (even with `--force`). If I add authentication to test the above, no changes are made on the backend. The API is wide open.

**Comment by josefaidt** [comment]
Hey @orome if you remove the restrictions with `amplify update api` and push, are you seeing the change reflected in the API Gateway console? for what it's worth here is an example repo with this setup https://github.com/josefaidt/amplify-restapi-cognito/blob/main/amplify/backend/api/apidbe4f741/override.ts

Please also consider adding a 👍 to this feature request for REST API + Cognito support https://github.com/aws-amplify/amplify-category-api/issues/345

**Comment by orome** [comment]
@josefaidt Sorry, Amplify is simply unusable in its current state. Here's the latest example:

Once I follow these instructions and override, it is **no longer possible to add any paths to the API** using `update api`. I can appear to do so, going through all the steps in the CLI and pushing without any errors, but the added path does not appear in API Gateway and is not available for invocation.
